concurrency-limits-core/src/main/java/com/netflix/concurrency/limits/limit/window/ImmutablePercentileSampleWindow.java

```diff
@@ -50,16 +49,13 @@
     }
 
     @Override
-    public ImmutablePercentileSampleWindow addSample(long rtt, int inflight) {
-        // TODO: very naive
-        // full copy in order to fulfill side-effect-free requirement of AtomicReference::updateAndGet
-        List<Long> newObservedRtts = new ArrayList<>(observedRtts);
-        newObservedRtts.add(rtt);
+    public ImmutablePercentileSampleWindow addSample(long rtt, long seqId, int inflight) {
+        observedRtts.add(new ObservedRtt(rtt, seqId));
```

Comment:
There is a race here that will result in a concurrency modification exception.  Set is no safer than List from concurrent modification.  

One way to address this is to have a fixed size array that is passed from window to window along with the number of samples in the array.  The array should be an AtomicLongArray to ensure updates are atomic.  Whenever addSample is called just set the next element of the array and increment the element counter.  It's OK if there's a race in updating the next element as such races can be considered noise.

Reply:
> There is a race here that will result in a concurrency modification exception. Set is no safer than List from concurrent modification.

Does it? It is just a series of additions, no delete operations occur, so ConcurrentModificationException should not happen. But yes, we can have inconsistency between reads and writes, and even with ConcurrentHashSet we can take into account some border samples twice.

Reply:
fixed here https://github.com/Netflix/concurrency-limits/pull/119/commits/49578124c291382fe0f7ccd4e8fdeb99adc580ab